src/main/java/org/interledger/ilp/exceptions/InterledgerException.java

```diff
@@ -1,36 +1,207 @@
-   */
-  public InterledgerException(String msg, Throwable cause) {
-    super(msg, cause);
-  }
-}
+public class InterledgerException extends RuntimeException {
+
+    private static final long serialVersionUID = 1L;
+
+    // TODO:(0) Add encoding support to serialize it through the wire (binary, JSON,...?, check the RFCs)
+    public enum ErrorType {
+        FINAL('F'),
+        TEMPORARY('T'),
+        RELATIVE('R');
+
+        private final String t;
+        private ErrorType(char t){
+            this.t = Character.toString(t);
+        }
+
+        @Override
+        public String toString(){
+            return t;
+        }
+    }
+
+    public enum ErrorCode {
+        // FINAL ERRORS            
+        F00_BAD_REQUEST               ("F00", "BAD REQUEST"             ),
+        F01_INVALID_PAQUET            ("F01", "INVALID PAQUET"          ),
+        F02_UNREACHABLE               ("F02", "UNREACHABLE"             ),
+        F03_INVALID_AMOUNT            ("F03", "INVALID AMOUNT"          ),
+        F04_INSUFFICIENT_DST_AMOUNT   ("F04", "INSUFFICIENT DST. AMOUNT"),
+        // TODO:(RFC) WRONG CONDITION is actually WRONG FULFILLMENT?
+        F05_WRONG_CONDITION           ("F05", "WRONG CONDITION"         ),
+        F06_UNEXPECTED_PAYMENT        ("F06", "UNEXPECTED PAYMENT"      ),
+        F07_CANNOT_RECEIVE            ("F07", "CANNOT RECEIVE"          ),
+        F99_APPLICATION_ERROR         ("F99", "APPLICATION ERROR"       ),
+
+        T00_INTERNAL_ERROR            ("T00", "INTERNAL ERROR"          ),
+        T01_LEDGER_UNREACHABLE        ("T01", "LEDGER UNREACHABLE"      ),
+        T02_LEDGER_BUSY               ("T02", "LEDGER BUSY"             ),
+        T03_CONNECTOR_BUSY            ("T03", "CONNECTOR BUSY"          ),
+        T04_INSUFFICIENT_LIQUIDITY    ("T04", "INSUFFICIENT LIQUIDITY"  ),
+        T05_RATE_LIMITED              ("T05", "RATE LIMITED"            ),
+        T99_APPLICATION_ERROR         ("T99", "APPLICATION ERROR"       ),
+
+        R00_TRANSFER_TIMED_OUT        ("R00", "TRANSFER TIMED OUT"      ),
+        R01_INSUFFICEINT_SOURCE_AMOUNT("R01", "INSUFFICEINT SOURCE AMOUNT"),
+        R02_INSUFFICIENT_TIMEOUT      ("R02", "INSUFFICIENT TIMEOUT"    ),
+        R99_APPLICATION_ERROR         ("R99", "APPLICATION ERROR"       )
+        ;
+        // TODO:(RFC) ADD FORBIDEN ERROR IF NO CREDENTIALS ARE PROVIDED OR LAUNCH "T01 LEDGER-UNREACHABLE" ???
+        // IMAGINE node1 -> conector1 -> conector2 -> (ERR. FORBIDEN) contector3
+        //  - How must routing be updated in  conector1/2/3 ?
+
+        public  final ErrorType errorType;
+        private final String code;
+        private final String name;
+
+        private ErrorCode(String code, String name) {
+            if (code==null || name==null){
+                throw new RuntimeException("code and/or name can not be null");
+            }
+            code = code.trim();
+            name = name.trim();
+            if (code.length()!=3) {
```

Comment:
I think this should only deal with Interledger Error codes which will always be 3 chars. Custom/application errors will need to extend this or have this as a `cause`.